adapters/silverpush/silverpush.go

```diff
@@ -0,0 +1,333 @@
+	BidFloor    float64 `json:"bidfloor"`
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	imps := request.Imp
+	var errors []error
+	requests := make([]*adapters.RequestData, 0, len(imps))
+
+	for _, imp := range imps {
+		impsByMediaType := impressionByMediaType(&imp)
+
+		request.Imp = []openrtb2.Imp{impsByMediaType}
+
+		if err := validateRequest(request); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		requestData, err := a.makeRequest(request)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		requests = append(requests, requestData)
+
+	}
+	return requests, errors
+}
+
+func (a *adapter) makeRequest(req *openrtb2.BidRequest) (*adapters.RequestData, error) {
+	reqJSON, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+	}, nil
+}
+
+func validateRequest(req *openrtb2.BidRequest) error {
+	imp := &req.Imp[0]
+	var silverPushExt openrtb_ext.ImpExtSilverpush
+	if err := setPublisherId(req, imp, &silverPushExt); err != nil {
+		return err
+	}
+	if err := setUser(req); err != nil {
+		return err
+	}
+
+	setDevice(req)
+
+	if err := setExtToRequest(req, silverPushExt.PublisherId); err != nil {
+		return err
+	}
+	return setImpForAdExchange(imp, &silverPushExt)
+}
+
+func setDevice(req *openrtb2.BidRequest) {
+
+	if req.Device == nil {
+		return
+	}
+	deviceCopy := *req.Device
+	if len(deviceCopy.UA) == 0 {
+		return
+	}
+	deviceCopy.OS = getOS(deviceCopy.UA)
+	if isMobile(deviceCopy.UA) {
+		deviceCopy.DeviceType = 1
+	} else if isCTV(deviceCopy.UA) {
+		deviceCopy.DeviceType = 3
+	} else {
+		deviceCopy.DeviceType = 2
+	}
+
+	req.Device = &deviceCopy
+}
+
+func setUser(req *openrtb2.BidRequest) error {
+	var extUser openrtb_ext.ExtUser
+	var userExtRaw map[string]json.RawMessage
+
+	if req.User != nil && req.User.Ext != nil {
+		if err := json.Unmarshal(req.User.Ext, &userExtRaw); err != nil {
+			return &errortypes.BadInput{Message: "Invalid user.ext."}
+		}
+		if userExtDataRaw, ok := userExtRaw["data"]; ok {
+			if err := json.Unmarshal(userExtDataRaw, &extUser); err != nil {
+				return &errortypes.BadInput{Message: "Invalid user.ext.data."}
+			}
+			var userCopy = *req.User
+			if isValidEids(extUser.Eids) {
+				userExt, err := json.Marshal(
+					&openrtb2.User{
+						EIDs: extUser.Eids,
+					})
+				if err != nil {
+					return &errortypes.BadInput{Message: "Error in marshaling user.eids."}
+				}
+
+				userCopy.Ext = userExt
+				req.User = &userCopy
+			}
+		}
+	}
+	return nil
+}
+
+func setExtToRequest(req *openrtb2.BidRequest, publisherID string) error {
+
```

Comment:
nitpick - Please remove this blank line.

Reply:
Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/41daf0668c923b8c72d21af84ad06a0f5fc47a62)